packages/composite-checkout-wpcom/src/hooks/use-shopping-cart.ts

```diff
@@ -392,14 +401,126 @@ export function useShoppingCart(
 	const shouldShowNotification = hookState.shouldShowNotification;
 
 	// Asynchronously initialize the cart. This should happen exactly once.
+	useInitializeCartFromServer(
+		cacheStatus,
+		canInitializeCart,
+		productToAdd,
+		getServerCart,
+		setServerCart,
+		hookDispatch,
+		onEvent
+	);
+
+	// Asynchronously re-validate when the cache is dirty.
+	useCartUpdateAndRevalidate( cacheStatus, responseCart, setServerCart, hookDispatch, onEvent );
+
+	// Keep a separate cache of the displayed cart which we regenerate only when
+	// the cart has been downloaded
+	const responseCartToDisplay = useCachedValidCart( cacheStatus, responseCart );
+
+	// Translate the responseCart into the format needed in checkout.
+	const cart: WPCOMCart = useMemo(
+		() => translateWpcomCartToCheckoutCart( translate, responseCartToDisplay ),
+		[ translate, responseCartToDisplay ]
+	);
+
+	useShowAddCouponSuccessMessage(
+		shouldShowNotification.didAddCoupon,
+		responseCart,
+		showAddCouponSuccessMessage,
+		hookDispatch
+	);
+
+	const addItem: ( ResponseCartProduct ) => void = useCallback( responseCartProductToAdd => {
+		hookDispatch( { type: 'ADD_CART_ITEM', responseCartProductToAdd } );
+	}, [] );
+
+	const removeItem: ( string ) => void = useCallback( uuidToRemove => {
+		hookDispatch( { type: 'REMOVE_CART_ITEM', uuidToRemove } );
+	}, [] );
+
+	const changeItemVariant: (
+		uuidToReplace: string,
+		newProductSlug: string,
+		newProductId: number
+	) => void = useCallback( ( uuidToReplace, newProductSlug, newProductId ) => {
+		hookDispatch( { type: 'REPLACE_CART_ITEM', uuidToReplace, newProductSlug, newProductId } );
+	}, [] );
+
+	const updateLocation: ( CartLocation ) => void = useCallback( location => {
+		hookDispatch( { type: 'SET_LOCATION', location } );
+	}, [] );
+
+	const submitCoupon: ( string ) => void = useCallback( couponToAdd => {
+		hookDispatch( { type: 'ADD_COUPON', couponToAdd } );
+	}, [] );
+
+	return {
+		isLoading: cacheStatus === 'fresh',
+		items: cart.items,
+		tax: cart.tax,
+		couponItem: cart.coupon,
+		total: cart.total,
+		subtotal: cart.subtotal,
+		credits: cart.credits,
+		errors: responseCart.messages?.errors ?? [],
+		allowedPaymentMethods: cart.allowedPaymentMethods,
+		addItem,
+		removeItem,
+		updateLocation,
+		changeItemVariant,
+		submitCoupon,
+		couponStatus,
+		couponCode: cart.couponCode,
+		variantRequestStatus,
+		variantSelectOverride,
+	} as ShoppingCartManager;
+}
+
+function useInitializeCartFromServer(
+	cacheStatus: CacheStatus,
+	canInitializeCart: boolean,
+	productToAdd: ResponseCartProduct | null,
+	getServerCart: Function,
```

Comment:
Ideally we could give a more specific type here, but because `getServerCart` is wrapped in a `useCallback` I'm not sure what it should be off the top of my head.

Reply:
Good question. [This post](https://fettblog.eu/typescript-react/hooks/#usememo---usecallback) seems to imply that we can use the type of the callback function. I suspect it will work because my guess is that `useCallback` is just a `useRef` under the hood, and `useRef` simply preserves an object reference across multiple renders. I'll try it here.